### PR TITLE
[DS-187] Improving Preprocessor Accuracy for Mnemonic Reader

### PR DESCRIPTION
**Pull request by happilyeverafter95**
## Goals and Notes for Reviewers

The goal of this PR is to improve the accuracy of Mnemonic Reader by creating a Version 2 Preprocessor which is more accurate. Comparison of Version 1 and Version 2 of the preprocessor can be found here: https://kiitehq.atlassian.net/wiki/spaces/PROD/pages/217743361/Preprocessing+Mnemonic+Reader+using+Latent+Semantic+Indexing

## Changes in the PR 

The preprocessor uses a similarity score to decide which file/paragraph(s) is the most likely to contain the ground truth answer. This PR adjusts this score.

- Previous Version 1 score was the cosine similarity between the question and document. Version 2 score multiples the cosine similarity with the intersectional score which is calculated by the % of similar words between the query and document

### CODE REVIEW

File: script/interactive.py [hunk 1]
```diff
@@ -160,7 +160,8 @@ def preprocess(text, question,n):
     for i in range(0, length):
         vectorizedParagraph = vectorizer.transform([processedParagraphs[i]])
         paragraph = svdModel.transform(vectorizedParagraph)
-        sim = cosine_similarity(processedQuestion, paragraph)
+        overlap = sum([1 if word in processedParagraphs[i] else 0 for word in tokenizedQuestion.split()])
```
Comment (Labibahx) [inline script/interactive.py]: so I see the point here is to give more weight to specific word overlap but is not that additional factor what the cos sim is calculating anyway

File: script/interactive.py [hunk 2]
```diff
@@ -239,22 +240,21 @@ def testPreprocessor(questions, n1, n2, doc = elvis):
     naiveScore = 0
     ignoreScore = 0 
     ignoreError = 0
-    cosine = []
     for i in range(0,n):
         correctParagraph = questions[i][2]
         ignore = questions[i][1]
```
Comment (Labibahx) [inline script/interactive.py]: is this "ignore" part of the questions?

File: script/interactive.py [hunk 3]
```diff
@@ -239,22 +240,21 @@ def testPreprocessor(questions, n1, n2, doc = elvis):
...
         result = processQuestion(questions[i][0], doc, n1, n2)[0]
         topParagraphs = [i[0] for i in result]
         if correctParagraph in topParagraphs[:n1]:
             naiveScore += 1
-        confidence = result[0][1][0][0]
-        if confidence <= 0.05 and ignore == True: 
+        topFile = result[0]
+        # the first index is for accessing the score, the second 2 are for unravelling the array
+        confidence = topFile[1][0][0]
+        if confidence <= 0.5 and ignore == True:
```
Comment (Labibahx) [inline script/interactive.py]: any reason why we are using 0.5 for the confidence?

File: script/interactive.py (hunk 1 above)
Comment (happilyeverafter95) [reply script/interactive.py]: To reiterate what was discussed: the word intersectionality score further strengthens the relationship between the question and the document(s). It bolstered preprocessor accuracy by 7-8% which was the reason why I decided to make this change

File: script/interactive.py (hunk 2 above)
Comment (happilyeverafter95) [reply script/interactive.py]: To reiterate what was discussed: the format of a question is (question, ignore, location) where ignore is a boolean value indicating whether the question has an answer that can be found in the documents.

File: script/interactive.py (hunk 3 above)
Comment (happilyeverafter95) [reply script/interactive.py]: To reiterate what was discussed: this is a placeholder value as we currently have no mechanisms in place to determine when Mnemonic Reader should answer a question.

**Pull request merged**